<center><h1>Z604 Assignment-3 Solutions : Parth Patel (pp20)</h1></center>

<h2>Importing the Reuters Dataset into a SFrame object</h2>

In [1]:
import os
import graphlab
sf = graphlab.SFrame.read_csv('/Users/Parth/Desktop/Z604BigDataLab/assignment3/reuters.csv', header=False)

2016-04-19 20:53:13,705 [INFO] graphlab.cython.cy_server, 176: GraphLab Create v1.8.5 started. Logging: /tmp/graphlab_server_1461113591.log


Finished parsing file /Users/Parth/Desktop/Z604BigDataLab/assignment3/reuters.csv

Parsing completed. Parsed 100 lines in 0.083798 secs.

This trial license of GraphLab Create is assigned to pp20@umail.iu.edu and will expire on April 28, 2016. Please contact trial@dato.com for licensing options or to request a free non-commercial license for personal or academic use.
------------------------------------------------------

Finished parsing file /Users/Parth/Desktop/Z604BigDataLab/assignment3/reuters.csv

Parsing completed. Parsed 2286 lines in 0.093439 secs.


Inferred types from first line of file as 
column_type_hints=[str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


<h2>Displaying Contents of SFrame</h2>

In [2]:
sf


Columns:
	X1	str
	X2	str

Rows: 2286

Data:
+-----+-------------------------------+
|  X1 |               X2              |
+-----+-------------------------------+
| ipi | French industrial producti... |
| ipi | UK Industrial production r... |
| ipi | Japans preliminary industr... |
| ipi | Japans industrial producti... |
| ipi | The Soviet economy recover... |
| ipi | British manufacturers expe... |
| ipi | Industrial output in Janua... |
| ipi | The growth rate of Brazili... |
| ipi | Swiss industrial output ro... |
| ipi | Chinas industrial output r... |
+-----+-------------------------------+
[2286 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

<h2>Task-2: Generating Feature Space for each Document</h2>

In [3]:
sf['bow'] = graphlab.text_analytics.count_words(sf['X2'])


In [4]:
sf['tfidf'] = graphlab.text_analytics.tf_idf(sf['bow'])


<h2>Task-1: Text Cleaning</h2>

In [5]:
docs = sf['bow'].dict_trim_by_values(2)


In [6]:
docs = docs.dict_trim_by_keys(graphlab.text_analytics.stopwords(), exclude=True)


<h2>Displaying Contents of SFrame with two new features.</h2>

In [7]:
sf

Columns:
	X1	str
	X2	str
	bow	dict
	tfidf	dict

Rows: 2286

Data:
+-----+-------------------------------+-------------------------------+
|  X1 |               X2              |              bow              |
+-----+-------------------------------+-------------------------------+
| ipi | French industrial producti... | {'and': 1, 'adjusted': 1, ... |
| ipi | UK Industrial production r... | {'reuter': 1, 'office': 2,... |
| ipi | Japans preliminary industr... | {'and': 1, 'adjusted': 3, ... |
| ipi | Japans industrial producti... | {'adjusted': 3, 'month': 2... |
| ipi | The Soviet economy recover... | {'sector': 2, 'all': 1, 's... |
| ipi | British manufacturers expe... | {'rating': 1, 'all': 1, 'r... |
| ipi | Industrial output in Janua... | {'and': 2, 'months': 3, 's... |
| ipi | The growth rate of Brazili... | {'and': 2, 'months': 3, 's... |
| ipi | Swiss industrial output ro... | {'and': 4, 'industrial': 1... |
| ipi | Chinas industrial output r... | {'and': 1, 'seven': 1, 'in... |
+-----+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'and': 0.1396744571382365... |
| {'reuter': 0.0070237338793... |
| {'and': 0.1396744571382365... |
| {'adjusted': 8.91715472867... |
| {'sector': 6.1803358904267... |
| {'rating': 6.3482644832348... |
| {'and': 0.2793489142764731... |
| {'and': 0.2793489142764731... |
| {'and': 0.5586978285529463... |
| {'and': 0.1396744571382365... |
+-------------------------------+
[2286 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

<h2>Ramdomly Splitting Data as Train (80%) and Test (20%)</h2>

In [23]:
train_data, test_data = sf.random_split(0.8)


<h2>Task-3: Using Different Classifiers to Classify Data</h2>

<h2>SVM Classifier</h2>

In [9]:
svm_combined_model = graphlab.svm_classifier.create(train_data, target='X1',
                       features = ['bow','tfidf'])

WARNING: The number of feature dimensions in this problem is very large in comparison with the number of examples. Unless an appropriate regularization value is set, this model may not provide accurate predictions for a validation/test set.

2016-04-19 20:57:54,311 [ERROR] graphlab.toolkits._main, 66: Toolkit error: Linear SVM currently only supports binary classification. Use the boosted_trees_classifier for multi-class classification.



PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



ToolkitError: Linear SVM currently only supports binary classification. Use the boosted_trees_classifier for multi-class classification.


<h2>Nearest Neighbor Classifier</h2>

In [10]:
nearestN_bow_model = graphlab.nearest_neighbor_classifier.create(train_data, target='X1',
                                    features = ['bow'])
nearestN_bow_results = nearestN_bow_model.evaluate(test_data)
print nearestN_bow_results

Starting brute force nearest neighbors model training.

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 465     | 0.0549149   | 8.634ms      |

| Done         |         | 100         | 338.507ms    |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 465     | 0.0549149   | 11.803ms     |

| Done         |         | 100         | 297.066ms    |

+--------------+---------+-------------+--------------+

{'confusion_matrix': Columns:
	target_label	str
	predicted_label	str
	count	int

Rows: 126

Data:
+-----------------+------------------+-------+
|   target_label  | predicted_label  | count |
+-----------------+------------------+-------+
|   crudenat-gas  |   crudenat-gas   |   8   |
|   money-fxdlr   |   money-fxdlr    |   22  |
|      copper     |    graincorn     |   1   |
| veg-oilpalm-oil | veg-oilpalm-oil  |   1   |
|      grain      |       gnp        |   1   |
|     nat-gas     |   crudenat-gas   |   4   |
|       gas       |       alum       |   1   |
|       gas       |      copper      |   1   |
|   crudenat-gas  |       jobs       |   1   |
|   money-supply  | interestmoney-fx |   1   |
+-----------------+------------------+-------+
[126 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns., 'accuracy': 0.6064516129032258}


In [24]:
nearestN_tfidf_model = graphlab.nearest_neighbor_classifier.create(train_data, target='X1',
                                    features = ['tfidf'])
nearestN_tfidf_results = nearestN_tfidf_model.evaluate(test_data)
print nearestN_tfidf_results

Starting brute force nearest neighbors model training.

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 478     | 0.0553097   | 11.481ms     |

| Done         |         | 100         | 504.337ms    |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 478     | 0.0553097   | 13.432ms     |

| Done         |         | 100         | 391.708ms    |

+--------------+---------+-------------+--------------+

{'confusion_matrix': Columns:
	target_label	str
	predicted_label	str
	count	int

Rows: 98

Data:
+------------------+------------------+-------+
|   target_label   | predicted_label  | count |
+------------------+------------------+-------+
|     tradebop     |     tradebop     |   1   |
|   money-supply   |   money-supply   |   20  |
|   money-fxdlr    |   money-fxdlr    |   28  |
|       ipi        |       cpi        |   2   |
|   crudenat-gas   |      grain       |   1   |
| livestockcarcass |     veg-oil      |   3   |
|       gas        |   crudenat-gas   |   1   |
|   crudenat-gas   |     nat-gas      |   1   |
|  oilseedsoybean  |     veg-oil      |   2   |
| livestockcarcass | livestockcarcass |   1   |
+------------------+------------------+-------+
[98 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns., 'accuracy': 0.6548117154811716}


In [13]:
nearestN_combined_model = graphlab.nearest_neighbor_classifier.create(train_data, target='X1',
                                    features = ['bow','tfidf'])
nearestN_combined_results = nearestN_combined_model.evaluate(test_data)
print nearestN_combined_results

Starting brute force nearest neighbors model training.

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 465     | 0.0549149   | 25.894ms     |

| Done         |         | 100         | 701.707ms    |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 465     | 0.0549149   | 28.451ms     |

| Done         |         | 100         | 590.193ms    |

+--------------+---------+-------------+--------------+

{'confusion_matrix': Columns:
	target_label	str
	predicted_label	str
	count	int

Rows: 106

Data:
+--------------+-----------------+-------+
| target_label | predicted_label | count |
+--------------+-----------------+-------+
|   pet-chem   |      grain      |   1   |
|    grain     |    graincorn    |   17  |
| money-fxdlr  |   money-fxdlr   |   22  |
| crudenat-gas |   crudenat-gas  |   10  |
|     bop      |       bop       |   8   |
|   veg-oil    |     veg-oil     |   6   |
|    copper    |      rubber     |   1   |
|    copper    |      copper     |   7   |
|     gas      |       gas       |   4   |
|    rubber    |      rubber     |   8   |
+--------------+-----------------+-------+
[106 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns., 'accuracy': 0.6516129032258065}


In [14]:
confusion_matrix = nearestN_combined_results.get('confusion_matrix')

targets = graphlab.SArray(confusion_matrix['target_label'])
predictions = graphlab.SArray(confusion_matrix['predicted_label'])

precision_score   = graphlab.evaluation.precision(targets, predictions)
recall_score  = graphlab.evaluation.recall(targets, predictions)

print "Precision of Nearest Neighbors Classifier: " + str(precision_score)
print "Recall of Nearest Neighbors Classifier: " + str(recall_score)

Precision of Nearest Neighbors Classifier: 0.392754736303
Recall of Nearest Neighbors Classifier: 0.295588235294


<h2>Decision Tree Classifier</h2>

In [25]:
decisionTree_model = graphlab.decision_tree_classifier.create(train_data, target='X1', max_depth = 5)
decisionTree_results = decisionTree_model.evaluate(test_data)
print decisionTree_results

WARNING: The number of feature dimensions in this problem is very large in comparison with the number of examples. Unless an appropriate regularization value is set, this model may not provide accurate predictions for a validation/test set.

WARNING: Detected extremely low variance for feature(s) 'tfidf' because all entries are nearly the same.
Proceeding with model training using all features. If the model does not provide results of adequate quality, exclude the above mentioned feature(s) from the input dataset.

Decision tree classifier:

--------------------------------------------------------

Number of examples          : 1703

Number of classes           : 34

Number of feature columns   : 3

Number of unpacked features : 34037

+-----------+--------------+-------------------+-------------------+---------------------+---------------------+

| Iteration | Elapsed Time | Training-accuracy | Training-log_loss | Validation-accuracy | Validation-log_loss |

+-----------+--------------+-------------------+-------------------+---------------------+---------------------+

| 1         | 1.049343     | 0.867880          | 1.293465          | 0.714286            | 1.724966            |

+-----------+--------------+-------------------+-------------------+---------------------+---------------------+

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.

{'f1_score': 0.6703676860176627, 'auc': 0.9339829558136492, 'recall': 0.6534578839077109, 'precision': 0.7258817960340894, 'log_loss': 1.8209153228118933, 'roc_curve': Columns:
	threshold	float
	fpr	float
	tpr	float
	p	int
	n	int
	class	int

Rows: 3400034

Data:
+-----------+-----+-----+---+-----+-------+
| threshold | fpr | tpr | p |  n  | class |
+-----------+-----+-----+---+-----+-------+
|    0.0    | 1.0 | 1.0 | 8 | 470 |   0   |
|   1e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   2e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   3e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   4e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   5e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   6e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   7e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   8e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   9e-05   | 1.0 | 1.0 | 8 | 470 |   

In [16]:
decisionTree_model.show(view="Tree")

Canvas is accessible via web browser at the URL: http://localhost:63705/index.html
Opening Canvas in default web browser.


<h2>Boosted Trees Classifier</h2>

In [26]:
boostedTree_model = graphlab.boosted_trees_classifier.create(train_data, target='X1',
                                 max_iterations=2, max_depth = 3)
boostedTree_results = boostedTree_model.evaluate(test_data)
print boostedTree_results

WARNING: The number of feature dimensions in this problem is very large in comparison with the number of examples. Unless an appropriate regularization value is set, this model may not provide accurate predictions for a validation/test set.

WARNING: Detected extremely low variance for feature(s) 'tfidf' because all entries are nearly the same.
Proceeding with model training using all features. If the model does not provide results of adequate quality, exclude the above mentioned feature(s) from the input dataset.

Boosted trees classifier:

--------------------------------------------------------

Number of examples          : 1729

Number of classes           : 34

Number of feature columns   : 3

Number of unpacked features : 34603

+-----------+--------------+-------------------+-------------------+---------------------+---------------------+

| Iteration | Elapsed Time | Training-accuracy | Training-log_loss | Validation-accuracy | Validation-log_loss |

+-----------+--------------+-------------------+-------------------+---------------------+---------------------+

| 1         | 0.734883     | 0.813765          | 1.436262          | 0.632911            | 1.743373            |

| 2         | 1.335478     | 0.829960          | 1.137638          | 0.645570            | 1.525460            |

+-----------+--------------+-------------------+-------------------+---------------------+---------------------+

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.

{'f1_score': 0.6956296504881868, 'auc': 0.9691643100936188, 'recall': 0.677073426381385, 'precision': 0.7794860874122103, 'log_loss': 1.5315525402976042, 'roc_curve': Columns:
	threshold	float
	fpr	float
	tpr	float
	p	int
	n	int
	class	int

Rows: 3400034

Data:
+-----------+-----+-----+---+-----+-------+
| threshold | fpr | tpr | p |  n  | class |
+-----------+-----+-----+---+-----+-------+
|    0.0    | 1.0 | 1.0 | 8 | 470 |   0   |
|   1e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   2e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   3e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   4e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   5e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   6e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   7e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   8e-05   | 1.0 | 1.0 | 8 | 470 |   0   |
|   9e-05   | 1.0 | 1.0 | 8 | 470 |   0

In [18]:
boostedTree_model.show(view="Tree")

Canvas is updated and available in a tab in the default browser.


<h2>Report</h2>

**1. Which classifier(s) can be more useful for this task? (hint: some classifiers can’t be used for this task)?**

As can be seen from the above file, I tried to implement all the four classifiers wiz., Support Vector Machines, Nearest Neighbor, Decision Tree and Boosted Decision Tree. All the models gave results except the SVM which is useful only when we are trying to do binary classification which is not the case with the Reuters dataset since it is a multi class dataset. 

**2. What kind of text representation (A.K.A features for the data) is more useful? Why?**

I tried using both the bag of words and the tfidf as features while building the model and found that the resuts using tfidf are better. TF-IDF returns documents that are highly relevant to a particular query rather than just returing all the documents with a threshold frequency, thus making it more accurate than the simple bag of words approach.

**3. Compare different classifiers' result by comparing model accuracy, auc, precision and recall (which one is the best)?**

The following results are taken from the evaluation task performed above using different classifiers and are as follows:

**Accuracy:**

SVM              - NA

Nearest Neighbor - 0.6516129032258065

Decision Tree    - 0.6882845188284519

Boosted Trees    - 0.7259414225941423


**AUC:**

SVM              - NA

Nearest Neighbor - nan

Decision Tree    - 0.9339829558136492

Boosted Trees    - 0.9691643100936188


**Precision:**

SVM              - NA

Nearest Neighbor - 0.392754736303

Decision Tree    - 0.725881796034

Boosted Trees    - 0.7794860874122


**Recall:**

SVM              - NA

Nearest Neighbor - 0.295588235294

Decision Tree    - 0.6534578839077109

Boosted Trees    - 0.677073426381385


The above results show that Boosted Decision trees provides us the best result in terms of accuracy. The major advantage of decision trees in general is that they are non parametric which means that they take care of the outliers by themselves. But the main disadvantage is that they easily overfit, boosted decision tress actually rectify this since they are ensemble methods over decision tress. Thus they eliminatet the need of any over tuning, giving better results.